### PR TITLE
Fix #218: Remove RuntimeMode mode parameter from Plugins

### LINKED ISSUE #218: Refactor Build mode
## Description

Does having a `kubernetes` build mode make sense when we have separate plugins (`openshift-maven-plugin` & `kubernetes-maven-plugin`)?

What's the point of having `mvn oc:build `-Djkube.mode=kubernetes` when we can always do `mvn k8s:build` even if we are running on an OpenShift cluster?

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fix #218 
Removed `jkube.mode` parameter from MOJOs. It won't be computed after this change(earlier default value was `RuntimeMode.auto`). Kubernetes Maven Plugin would send `RuntimeMode.kubernetes` and Openshift Maven Plugin would send `Runtimemode.openshift`.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have implemented unit tests to cover my changes
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
Quarkus tests for OpenShift will never pass since they are relying on [Kubernetes mode](https://github.com/jkubeio/jkube-integration-tests/blob/d666125c0a94f5b50c991f75eaed08ab7d19705f/it/src/test/java/org/eclipse/jkube/integrationtests/quarkus/rest/QuarkusOcITCase.java#L77) ([here too](https://github.com/jkubeio/jkube-integration-tests/blob/d666125c0a94f5b50c991f75eaed08ab7d19705f/it/src/test/java/org/eclipse/jkube/integrationtests/quarkus/rest/QuarkusOcITCase.java#L91)).

Try to update the tests so they use the `jkube.build.strategy=docker` property instead. Once tests are updated and merged, the full test suite should be successful.

**Comment by rohanKanojia** [comment]
ah, ok

File: jkube-kit/config/resource/src/main/java/org/eclipse/jkube/kit/config/resource/RuntimeMode.java [hunk 1]
```diff
@@ -33,36 +33,23 @@
      * onto cluster. It can be used both on vanilla Kubernetes and
      * OpenShift.
      */
-    kubernetes(false, "Kubernetes"),
+    KUBERNETES("Kubernetes"),
```
Comment (manusa) [inline jkube-kit/config/resource/src/main/java/org/eclipse/jkube/kit/config/resource/RuntimeMode.java:36]: Plexus deserialization is not configurable for enums (this is the reason why this values were still in lower-case).
Make sure that this value is never deserialized from a Maven property (i.e. `jkube.mode`), if it still is deserialized somewhere, it will fail).

File: jkube-kit/config/resource/src/main/java/org/eclipse/jkube/kit/config/resource/RuntimeMode.java [hunk 2]
```diff
@@ -33,36 +33,23 @@
...
+    OPENSHIFT("OpenShift");
 
-    /**
-     * Detect automatically whether running cluster is OpenShift or Kuberentes.
-     * This is done by contacting cluster API server.
-     */
-    auto(true, "Auto");
-
-    public static final RuntimeMode DEFAULT = RuntimeMode.auto;
     public static final String FABRIC8_EFFECTIVE_PLATFORM_MODE = "jkube.internal.effective.platform.mode";
```
Comment (manusa) [inline jkube-kit/config/resource/src/main/java/org/eclipse/jkube/kit/config/resource/RuntimeMode.java:45]: Maybe it's time to rename this constant

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/JKubeServiceHub.java [hunk 3]
```diff
@@ -71,16 +70,9 @@ public JKubeServiceHub(
     private void init() {
         Objects.requireNonNull(configuration, "JKubeKitConfiguration is required");
         Objects.requireNonNull(log, "log is a required parameter");
-        if (platformMode == null) {
```
Comment (manusa) [inline jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/JKubeServiceHub.java:74]: If platformMode is now required, add the nullcheck:
`Objects.requireNonNull(platformMode, "platformMode is a required parameter");`

File: jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/openshift/AutoTLSEnricherTest.java [hunk 4]
```diff
@@ -91,18 +91,18 @@ private AdaptTestConfig(PlatformMode mode, String initContainerNameConfig, Strin
...
-                new AdaptTestConfig(PlatformMode.openshift, null, "tls-jks-converter", null,
+                new AdaptTestConfig(RuntimeMode.OPENSHIFT, null, "tls-jks-converter", null,
                         "jimmidyson/pemtokeystore:v0.1.0", null, "tls-pem", "jks-b", "jks-b"),
-                new AdaptTestConfig(PlatformMode.openshift, "test-container-name", "test-container-name", "image/123",
+                new AdaptTestConfig(RuntimeMode.OPENSHIFT, "test-container-name", "test-container-name", "image/123",
                         "image/123", "tls-a", "tls-a", "jks-b", "jks-b") };
 
         for (final AdaptTestConfig tc : data) {
-            TreeMap configMap = new TreeMap() {
+            TreeMap<String, String> configMap = new TreeMap<String, String>() {
```
Comment (manusa) [inline jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/openshift/AutoTLSEnricherTest.java:105]:
```suggestion
            TreeMap<String, String> configMap = new TreeMap<>() {
```

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-false/invoker.properties [hunk 5]
```diff
@@ -13,5 +13,5 @@
 #
 
 invoker.goals.1=clean k8s:resource
-invoker.mavenOpts=-Djkube.verbose -Djkube.mode=openshift -Dfabric8.openshift.generateRoute=false
+invoker.mavenOpts=-Djkube.verbose -Dfabric8.openshift.generateRoute=false
```
Comment (manusa) [inline kubernetes-maven-plugin/it/src/it/simple-with-route-flag-false/invoker.properties:16]: Probably the second option `-Dfabric8.openshift.generateRoute=false` has no effect (if it does then something is wrong), we should remove it.

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/invoker.properties [hunk 6]
```diff
@@ -13,5 +13,5 @@
 #
 
 invoker.goals.1=clean k8s:resource
-invoker.mavenOpts=-Djkube.verbose -Djkube.mode=openshift -Dfabric8.openshift.generateRoute=true
-invoker.debug=false
\ No newline at end of file
+invoker.mavenOpts=-Djkube.verbose -Djkube.openshift.generateRoute=true
```
Comment (manusa) [inline kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/invoker.properties:16]: https://github.com/eclipse/jkube/pull/244/files#r442657988

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml [hunk 7]
```diff
@@ -84,7 +84,7 @@ items:
             valueFrom:
               fieldRef:
                 fieldPath: metadata.namespace
-          image: jkube-maven-sample-zero-config:latest
+          image: jkube/jkube-maven-sample-zero-config:latest
```
Comment (manusa) [inline kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: Is this because previously this file was ignored?

File: openshift-maven-plugin/it/src/it/simple-with-route-flag-false/invoker.properties [hunk 8]
```diff
@@ -13,5 +13,5 @@
 #
 
 invoker.goals.1=clean oc:resource
-invoker.mavenOpts=-Djkube.verbose -Djkube.mode=openshift -Dfabric8.openshift.generateRoute=false
+invoker.mavenOpts=-Djkube.verbose -Dfabric8.openshift.generateRoute=false
```
Comment (manusa) [inline openshift-maven-plugin/it/src/it/simple-with-route-flag-false/invoker.properties:16]: https://github.com/eclipse/jkube/pull/244/files#r442657988

Comment (manusa) [reply openshift-maven-plugin/it/src/it/simple-with-route-flag-false/invoker.properties:16]: Invalid comment, please ignore

File: openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java [hunk 9]
```diff
@@ -28,6 +30,48 @@
...
+    private Boolean generateRoute;
+
+    @Parameter(property = "jkube.openshift.enableAutomaticTrigger", defaultValue = "true")
+    private Boolean enableAutomaticTrigger;
+
+    @Parameter(property = "jkube.openshift.imageChangeTrigger", defaultValue = "true")
+    private Boolean enableImageChangeTrigger;
+
+    @Parameter(property = "jkube.openshift.enrichAllWithImageChangeTrigger", defaultValue = "false")
+    private Boolean erichAllWithImageChangeTrigger;
```
Comment (manusa) [inline openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java:73]: Are these properties dead? Could you open an issue providing details (if you know about them)?
It seems that at least some of them are used on Enrichers, but are not read from the Mojo.

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml (hunk 7 above)
Comment (rohanKanojia) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: Actually, earlier this test was running with `jkube.mode=openshift`. With this PR, runtime mode is KUBERNETES for Kubernetes Maven Plugin.
https://github.com/eclipse/jkube/blob/48ccd83fc8ffd81e3d8716b4fa8b43177ae6755d/kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/invoker.properties#L16

File: openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java (hunk 9 above)
Comment (rohanKanojia) [reply openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java:73]: Umm, you mean add javadocs. I see these are covered in plugin documentation: https://www.eclipse.org/jkube/docs/openshift-maven-plugin#Supported-Properties-Resource

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml (hunk 7 above)
Comment (manusa) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: So this file was being ignored previously, right? i.e. if we delete it from master/main branch, tests won't break

Comment (rohanKanojia) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: I don't think this file is ignored.

File: openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java (hunk 9 above)
Comment (manusa) [reply openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java:73]: Ok, let me rephrase. These properties are dead in the Mojo, they are directly computed/retrieved in the Enricher, meaning these fields can be deleted (fields are ignored, same as provided defaults).
So what we need is an issue providing details about what these properties do (link to docs is good), and what should we do with them. Basically, we need an answer to why are they in the Mojo if they are ignored > this seems like the result of an abandoned work in progress. And what we want, complete their movement to the Mojo (properties no longer retrieved from Enricher) **or** complete their movement to the Enricher (delete them from the Mojo).

Comment (rohanKanojia) [reply openshift-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/OpenshiftResourceMojo.java:73]: Thanks for explanation. I've filed https://github.com/eclipse/jkube/issues/259 for this

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml (hunk 7 above)
Comment (manusa) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: If this is the case, was previous behavior was wrong? Why were `kubernetes.yml` files generated if runtime mode was `openshift`?

Comment (rohanKanojia) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: I think `kubernetes.yml` were getting generated because `kubernetes-maven-plugin` goals were being invoked. This `jkube.mode` property was changing default generator image name depending upon runtime mode. Earlier with `jkube.mode` set to Openshift, they were generating image name for Openshift builds. Right now we have same test in openshift and it's image field looks like this:
https://github.com/eclipse/jkube/blob/b2e25f66df900beede673ad4d236f1f8e900eb0e/openshift-maven-plugin/it/src/it/simple-with-route-flag-true/expected/openshift.yml#L90

File: jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/openshift/AutoTLSEnricherTest.java (hunk 4 above)
Comment (rohanKanojia) [reply jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/openshift/AutoTLSEnricherTest.java]: I tried doing this but I couldn't use '<>' with anonymous inner classes. Our project is at JDK8 right now, maybe we should upgrade to JDK11? WDYT?

Comment (manusa) [reply jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/openshift/AutoTLSEnricherTest.java:105]: We should not be doing this, this is really bad.
Creating an inner class extending a TreeMap is something I don't see really the purpose here, and is really a bad practice.
Just instantiate the Map and put the entries after.

File: kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml (hunk 7 above)
Comment (manusa) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: > Right now we have same test in openshift and it's image field looks like this:

Yes, but it's generated in an `openshift.yml` file (as expected).

The fact that this wasn't the case before, and that instead of generating an `openshift.yml` file a `kubernetes.yml` file was being generated with openshift specific configurations (i.e. `image`)  and not even completed with deploymentconfig files, makes me understand that previous behavior was wrong.

So conclusion is that this file was not being ignored before due to an inconsistent behavior, but after the current PR, things will behave as expected and only specific cluster files will be generated with the **right** configs.

Comment (rohanKanojia) [reply kubernetes-maven-plugin/it/src/it/simple-with-route-flag-true/expected/kubernetes.yml:87]: cool, thanks. I would mark this as resolved.
